src/Processors/Transforms/MergingAggregatedTransform.cpp

```diff
@@ -10,11 +11,118 @@ namespace ErrorCodes
     extern const int LOGICAL_ERROR;
 }
 
+Block MergingAggregatedTransform::appendGroupingIfNeeded(const Block & in_header, Block out_header)
+{
+    /// __grouping_set is neither GROUP BY key nor an aggregate function.
+    /// It behaves like a GROUP BY key, but we cannot append it to keys
+    /// because it changes hashing method and buckets for two level aggregation.
+    /// Now, this column is processed "manually" by merging each group separately.
+    if (in_header.has("__grouping_set"))
+        out_header.insert(0, in_header.getByName("__grouping_set"));
+
+    return out_header;
+}
+
 MergingAggregatedTransform::MergingAggregatedTransform(
     Block header_, AggregatingTransformParamsPtr params_, size_t max_threads_)
-    : IAccumulatingTransform(std::move(header_), params_->getHeader())
-    , params(std::move(params_)), max_threads(max_threads_)
+    : IAccumulatingTransform(header_, appendGroupingIfNeeded(header_, params_->getHeader()))
+    , params(std::move(params_)), max_threads(max_threads_), has_grouping_sets(header_.has("__grouping_set"))
+{
+}
+
+void MergingAggregatedTransform::addBlock(Block block)
+{
+    if (!has_grouping_sets)
+    {
+        auto & bucket_to_blocks = grouping_sets[0];
+        bucket_to_blocks[block.info.bucket_num].emplace_back(std::move(block));
+        return;
+    }
+
+    auto grouping_position = block.getPositionByName("__grouping_set");
+    auto grouping_column = block.getByPosition(grouping_position).column;
+    block.erase(grouping_position);
+
+    /// Split a block by __grouping_set values.
+
+    const auto * grouping_column_typed = typeid_cast<const ColumnUInt64 *>(grouping_column.get());
+    if (!grouping_column_typed)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Expected UInt64 column for __grouping_set, got {}", grouping_column->getName());
+
+    /// Enumerate groups and fill the selector.
+    std::map<UInt64, size_t> enumerated_groups;
+    IColumn::Selector selector;
+
+    const auto & grouping_data = grouping_column_typed->getData();
+    size_t num_rows = grouping_data.size();
+    UInt64 last_group = grouping_data[0];
+    for (size_t row = 1; row < num_rows; ++row)
+    {
+        auto group = grouping_data[row];
+
+        /// Optimization for equal ranges.
+        if (last_group == group)
+            continue;
+
+        /// Optimization for single group.
+        if (enumerated_groups.empty())
+        {
+            selector.reserve(num_rows);
+            enumerated_groups.emplace(last_group, enumerated_groups.size());
```

Comment:
This line looks redundant because there's another `emplace` call outside of `if`

Reply:
Ah, yes, there is a bug below.